lib/cloud/clients.go

```diff
@@ -942,3 +1029,108 @@ func (c *TestCloudClients) GetAzureRunCommandClient(subscription string) (azure.
 func (c *TestCloudClients) Close() error {
 	return nil
 }
+
+// AWSSessionCacheKeyBuilder is a helper struct wrapper around a string builder
+// that is used to efficiently build cache keys for a given region and, optionally,
+// a chain of AWS assumed-role sessions.
+type AWSSessionCacheKeyBuilder struct {
+	sb        *strings.Builder
+	roleCount int
+}
+
+// NewAWSSessionCacheKeyBuilder constructs a new cache key builder intialized
+// with a specific AWS region.
+func NewAWSSessionCacheKeyBuilder(region string) AWSSessionCacheKeyBuilder {
+	sb := strings.Builder{}
+	_, _ = sb.WriteString(region) // infalliable.
+	return AWSSessionCacheKeyBuilder{sb: &sb}
+}
+
+// AddRole adds an AWS IAM role to the cache key builder.
+func (c *AWSSessionCacheKeyBuilder) AddRole(awsRole services.AssumeRole) {
+	// Role ARN can potentially come from user input via --db-user for some
+	// protocols.
+	// Out of paranoia for crafted user input, I build the key in a way that
+	// will avoid fetching an AWS session from cache that the user should not
+	// have access to.
+	key := fmt.Sprintf(":Role[%d]:ARN[%v]:ExternalID[%v]", c.roleCount, awsRole.RoleARN, awsRole.ExternalID)
+	_, _ = c.sb.WriteString(key) // infalliable.
+	c.roleCount++
+}
+
+func (c *AWSSessionCacheKeyBuilder) String() string {
+	return c.sb.String()
+}
+
+// checkAndSetAssumeRoles checks a chain of AWS IAM roles for obvious errors that
+// would return an error from AWS API, and eliminates superfluous external IDs
+// from the chain.
+// "obvious" errors include:
+//   - ARN fails to parse, or is not an IAM Role.
+//   - a role is in a different partition than the given region.
+//   - a role is in a different account than the next role and the next role
+//     does not have an external ID.
+//
+// "superfluous external IDs" are those for a role that is in the same account
+// as the role preceding it in the chain.
+func checkAndSetAssumeRoles(region string, roles []services.AssumeRole) ([]services.AssumeRole, error) {
```

Comment:
this function is used only in test. We should split the implementation from the helper tests function.

Reply:
I don't think it's worth enforcing these checks here anymore. Removed this and its tests in fc800e15f75f7ed8f38b88b6aba00a756c66d144